File: zombie-survivor/python/src/zombie_survivor/survivor.py
from .equipment import Equipment, EquipmentSlot
from .exceptions import EquipmentCapacityExceededException
from .level import Level, level_for
from .skill import Skill

_BASE_ACTIONS = 3
_MAX_WOUNDS = 2
_BASE_CAPACITY = 5
_MAX_IN_HAND = 2


class Survivor:
    def __init__(self, name: str) -> None:
        self._name = name
        self._wounds = 0
        self._experience = 0
        self._equipment: list[Equipment] = []
        self._skills: list[Skill] = []

    @property
    def name(self) -> str:
        return self._name

    @property
    def wounds(self) -> int:
        return self._wounds

    @property
    def experience(self) -> int:
        return self._experience

    @property
    def is_alive(self) -> bool:
        return self._wounds < _MAX_WOUNDS

    @property
    def level(self) -> Level:
        return level_for(self._experience)

    @property
    def actions_per_turn(self) -> int:
        bonus = sum(1 for s in self._skills if s == Skill.PLUS_ONE_ACTION)
        return _BASE_ACTIONS + bonus

    @property
    def equipment_capacity(self) -> int:
        hoard_bonus = sum(1 for s in self._skills if s == Skill.HOARD)
        return _BASE_CAPACITY - self._wounds + hoard_bonus

    @property
    def equipment(self) -> list[Equipment]:
        return list(self._equipment)

    @property
    def skills(self) -> list[Skill]:
        return list(self._skills)

    @property
    def in_hand_count(self) -> int:
        return sum(1 for e in self._equipment if e.slot == EquipmentSlot.IN_HAND)

    @property
    def in_reserve_count(self) -> int:
        return sum(1 for e in self._equipment if e.slot == EquipmentSlot.IN_RESERVE)

    def receive_wound(self) -> bool:
        if not self.is_alive:
            return False
        self._wounds += 1
        return True

    def equip(self, item_name: str) -> None:
        if len(self._equipment) >= self.equipment_capacity:
            raise EquipmentCapacityExceededException(
                f"Cannot carry more than {self.equipment_capacity} pieces of equipment."
            )
        slot = EquipmentSlot.IN_HAND if self.in_hand_count < _MAX_IN_HAND else EquipmentSlot.IN_RESERVE
        self._equipment.append(Equipment(item_name, slot))

    def discard(self, item_name: str) -> None:
        for i, e in enumerate(self._equipment):
            if e.name == item_name:
                self._equipment.pop(i)
                return

    @property
    def needs_to_discard(self) -> bool:
        return len(self._equipment) > self.equipment_capacity

    def kill_zombie(self) -> None:
        self._experience += 1

    def check_level_up(self, previous_experience: int) -> Level | None:
        previous_level = level_for(previous_experience)
        if self.level != previous_level:
            return self.level
        return None

    def unlock_skill(self, skill: Skill) -> None:
        self._skills.append(skill)

File: zombie-survivor/typescript/src/Survivor.ts
import type { Equipment } from './Equipment.js';
import { EquipmentSlot } from './Equipment.js';
import { EquipmentCapacityExceededException } from './Exceptions.js';
import { Level, levelFor } from './Level.js';
import { Skill } from './Skill.js';

const BaseActions = 3;
const MaxWounds = 2;
const BaseCapacity = 5;
const MaxInHand = 2;

export class Survivor {
  private readonly _equipment: Equipment[] = [];
  private readonly _skills: Skill[] = [];
  private _wounds = 0;
  private _experience = 0;

  constructor(public readonly name: string) {}

  get wounds(): number { return this._wounds; }
  get experience(): number { return this._experience; }
  get isAlive(): boolean { return this._wounds < MaxWounds; }

  get level(): Level { return levelFor(this._experience); }

  get actionsPerTurn(): number {
    const bonus = this._skills.filter(s => s === Skill.PlusOneAction).length;
    return BaseActions + bonus;
  }

  get equipmentCapacity(): number {
    const hoardBonus = this._skills.filter(s => s === Skill.Hoard).length;
    return BaseCapacity - this._wounds + hoardBonus;
  }

  get equipment(): readonly Equipment[] { return this._equipment; }
  get skills(): readonly Skill[] { return this._skills; }

  get inHandCount(): number {
    return this._equipment.filter(e => e.slot === EquipmentSlot.InHand).length;
  }

  get inReserveCount(): number {
    return this._equipment.filter(e => e.slot === EquipmentSlot.InReserve).length;
  }

  receiveWound(): boolean {
    if (!this.isAlive) return false;
    this._wounds++;
    return true;
  }

  equip(itemName: string): void {
    if (this._equipment.length >= this.equipmentCapacity) {
      throw new EquipmentCapacityExceededException(
        `Cannot carry more than ${this.equipmentCapacity} pieces of equipment.`,
      );
    }
    const slot = this.inHandCount < MaxInHand ? EquipmentSlot.InHand : EquipmentSlot.InReserve;
    this._equipment.push({ name: itemName, slot });
  }

  discard(itemName: string): void {
    const index = this._equipment.findIndex(e => e.name === itemName);
    if (index >= 0) this._equipment.splice(index, 1);
  }

  get needsToDiscard(): boolean {
    return this._equipment.length > this.equipmentCapacity;
  }

  killZombie(): void {
    this._experience++;
  }

  checkLevelUp(previousExperience: number): Level | null {
    const previousLevel = levelFor(previousExperience);
    if (this.level !== previousLevel) return this.level;
    return null;
  }

  unlockSkill(skill: Skill): void {
    this._skills.push(skill);
  }
}
